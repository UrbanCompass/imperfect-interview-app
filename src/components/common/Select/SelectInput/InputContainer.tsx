import React, { useContext } from 'react';
import styled from '@emotion/styled';

import { SelectContext, SelectContextProps } from '../Select';

export const Input = styled.input`
  box-sizing: border-box;
  user-select: text;
  width: 100%;
`;
Input.displayName = 'Input';

const InputWrapper = styled.div`
  display: inline-flex;
  align-items: center;
  width: 100%;
`;
InputWrapper.displayName = 'InputWrapper';

export function InputContainer(): JSX.Element {
  const {
    // downshift
    getInputProps,
    toggleMenu,
    // props
    autoFocus,
    inputId,
    placeholder,
    inputStyle,
    value,
    onValueChange,
  } = useContext(SelectContext) as SelectContextProps;

  return (
    <InputWrapper>
      <Input
        {...getInputProps({
          id: inputId,
          type: 'text',
          className: 'cx-selectField-arrow',
          placeholder,
          autoFocus,
          spellCheck: false,
          autoComplete: 'off',
          autoCorrect: 'off',
          autoCapitalize: 'off',
          style: inputStyle,
          onKeyDown: (event) => {
            /* istanbul ignore else */
            if (value && onValueChange && event.key === 'Enter') {
              onValueChange(value);
              const inputElement = document.getElementById(inputId);
              /* istanbul ignore next */
              if (inputElement) {
                inputElement.blur();
              }
            }
          },
        })}
        data-tn={`tours-${inputId}-select-input`}
        onClick={(): void => toggleMenu()}
      />
    </InputWrapper>
  );
}
