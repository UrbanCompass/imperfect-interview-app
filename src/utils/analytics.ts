import { ListingStatus, Media, MusicTrack, ProcessedListing } from '@/types';
import { SOURCE_APP } from '@/constants';

interface ListingProperties {
  listingId?: string;
  listingStatus?: ListingStatus;
  totalImagesCount: number;
}

function track(name: string, extraProperties = {}): void {
  const properties = {
    product: SOURCE_APP,
    ...extraProperties,
  };

  if (window.analytics) {
    window.analytics.track(name, properties);
  }
}

function getTotalImageCount(listing: ProcessedListing): number {
  return listing?.media?.length || 0;
}

function getSelectedImageCount(media: Media[]): number {
  return media.length;
}

function getUploadedImageCount(media: Media[]): number {
  const uploaded = media?.filter((m) => !!m.isCustomUpload);
  return uploaded.length;
}

function isSqftEnabled(media: Media[]): boolean {
  return !!media?.some((m) => m.showFootage);
}

function getCustomTextCount(media: Media[]): number {
  const withText = media?.filter((m) => !!m.note) || [];
  return withText.length;
}

function getListingAttributeCount(media: Media[]): number {
  const withListingAttribute = media?.filter((m) => !!m.textSource) || [];
  return withListingAttribute.length;
}

function isAudioEnabled(music: MusicTrack): boolean {
  return !!music;
}

function getTrackName(music: MusicTrack): string {
  return music?.name;
}

function getListingProperties(listing: ProcessedListing): ListingProperties {
  return {
    listingId: listing?.listingIdSHA,
    listingStatus: listing?.status,
    totalImagesCount: getTotalImageCount(listing),
  };
}

function getImageSelectionProperties(
  listing: ProcessedListing,
  media: Media[]
): Record<string, unknown> {
  return {
    ...getListingProperties(listing),
    selectedImagesCount: getSelectedImageCount(media),
    uploadedImagesCount: getUploadedImageCount(media),
  };
}

function getFullVideoProperties(
  listing: ProcessedListing,
  media: Media[],
  music: MusicTrack
): Record<string, unknown> {
  return {
    ...getImageSelectionProperties(listing, media),
    sqftEnabled: isSqftEnabled(media),
    customTextCount: getCustomTextCount(media),
    listingAttributeCount: getListingAttributeCount(media),
    audioEnabled: isAudioEnabled(music),
    audioTrackName: getTrackName(music),
  };
}

export function trackHomepageViewed(): void {
  track('Video Generator Viewed Homepage');
}

export function trackListingSelected(listingIdSHA: string): void {
  track('Video Generator Selected Listing', {
    listingId: listingIdSHA,
  });
}

export function trackAutoGenerateClicked(): void {
  track('Video Generator Clicked Auto Generate');
}

export function trackCreateCustomClicked(): void {
  track('Video Generator Clicked Create Custom');
}

export function trackCreateCustomBrandingClicked(): void {
  track('Video Generator Clicked Create Custom (Branding Video)');
}

export function trackImageSelectionPageViewed(listing, autoGenerated, videoUrl): void {
  track('Video Generator Viewed Image Selection Page', {
    ...getListingProperties(listing),
    autoGenerated,
    videoUrl,
  });
}

export function trackBrandingImageSelectionPageViewed(
  listingStatusFilter,
  autoGenerated,
  videoUrl
): void {
  track('Video Generator Viewed Image Selection Page (Branding Video)', {
    listingStatusFilter,
    autoGenerated,
    videoUrl,
  });
}

export function trackImagesUploaded(count: number, autoGenerated, videoUrl): void {
  track('Video Generator Uploaded Images', { count, autoGenerated, videoUrl });
}

export function trackImageSelected(autoGenerated, videoUrl): void {
  track('Video Generator Selected Image', { autoGenerated, videoUrl });
}

export function trackImageUnselected(autoGenerated, videoUrl): void {
  track('Video Generator Unselected Image', { autoGenerated, videoUrl });
}

export function trackImagesSelected(listing, media, autoGenerated, videoUrl): void {
  track('Video Generator Selected Images', {
    ...getImageSelectionProperties(listing, media),
    autoGenerated,
    videoUrl,
  });
}

export function trackCustomizationPageViewed(listing, media, autoGenerated, videoUrl): void {
  track('Video Generator Viewed Customization Page', {
    ...getImageSelectionProperties(listing, media),
    autoGenerated,
    videoUrl,
  });
}

export function trackImageReordered(autoGenerated, videoUrl): void {
  track('Video Generator Reordered Image', { autoGenerated, videoUrl });
}

export function trackImageRemoved(autoGenerated, videoUrl): void {
  track('Video Generator Removed Image', { autoGenerated, videoUrl });
}

export function trackAgentBrandingScreenRemoved(autoGenerated, videoUrl): void {
  track('Video Generator Removed Agent Branding Screen', { autoGenerated, videoUrl });
}

export function trackCompassLogoScreenRemoved(autoGenerated, videoUrl): void {
  track('Video Generator Removed Compass Logo Screen', { autoGenerated, videoUrl });
}

export function trackListingAttributeSelected(attribute, autoGenerated, videoUrl): void {
  track('Video Generator Selected Listing Attribute', {
    attribute,
    autoGenerated,
    videoUrl,
  });
}

export function trackAudioPlayed(music): void {
  track('Video Generator Played Audio', {
    audioTrackName: getTrackName(music),
  });
}

export function trackAudioPaused(music): void {
  track('Video Generator Paused Audio', {
    audioTrackName: getTrackName(music),
  });
}

export function trackAudioDisabled(autoGenerated, videoUrl): void {
  track('Video Generator Disabled Audio', {
    autoGenerated,
    videoUrl,
  });
}

export function trackAudioSelected(music, autoGenerated, videoUrl): void {
  track('Video Generator Selected Audio', {
    audioTrackName: getTrackName(music),
    autoGenerated,
    videoUrl,
  });
}

export function trackVideoCustomized(listing, media, music, autoGenerated, videoUrl): void {
  track('Video Generator Customized Video', {
    ...getFullVideoProperties(listing, media, music),
    autoGenerated,
    videoUrl,
  });
}

export function trackProgressPageViewed(listing, media, music, autoGenerated, videoUrl): void {
  track('Video Generator Viewed Progress Page', {
    ...getFullVideoProperties(listing, media, music),
    autoGenerated,
    videoUrl,
  });
}

export function trackDownloadPageViewed(listing, media, music, autoGenerated, videoUrl): void {
  track('Video Generator Viewed Download Page', {
    ...getFullVideoProperties(listing, media, music),
    autoGenerated,
    videoUrl,
  });
}

export function trackDownloadClicked(listing, media, music, autoGenerated, videoUrl): void {
  track('Video Generator Clicked Download', {
    ...getFullVideoProperties(listing, media, music),
    autoGenerated,
    videoUrl,
  });
}

export function trackStartOverClicked(autoGenerated, videoUrl): void {
  track('Video Generator Clicked Start Over', { autoGenerated, videoUrl });
}

export function trackEditVideoClicked(autoGenerated, videoUrl): void {
  track('Video Generator Clicked Edit Video', { autoGenerated, videoUrl });
}

export function trackBackToDigitalAdsClicked(): void {
  track('Video Generator Clicked Back To Digital Ads');
}

export function trackCreateVideoAdClicked(adType): void {
  track('Video Generator Clicked Create Video Ad', { adType });
}
