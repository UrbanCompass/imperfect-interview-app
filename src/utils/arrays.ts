/**
 * @param arr Any array
 * @param newItems An array of items to add, or a single non-array item
 * @returns A copy of `arr` with all items in `newItems` appended
 */
export function arrayWithItems<T>(arr: Array<T>, newItems: Array<T> | T): Array<T> {
  const arrCopy = arr.slice();
  return arrCopy.concat(newItems);
}

/**
 * @param arr An array
 * @param removeItems An array of items to remove, or a single non-array item
 * @returns A copy of `arr` with all items in `removeItems` removed
 */
export function arrayWithoutItems<T>(arr: Array<T>, removeItems: Array<T> | T): Array<T> {
  const arrCopy = arr.slice();
  const removeItemsArray = Array.isArray(removeItems) ? removeItems : [removeItems];
  removeItemsArray.forEach((removeItem) => {
    if (arrCopy.includes(removeItem)) {
      arrCopy.splice(arrCopy.indexOf(removeItem), 1);
    }
  });
  return arrCopy;
}
