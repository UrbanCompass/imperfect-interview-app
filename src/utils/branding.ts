import { listSearch } from '@/api';
import { NUM_IMAGES_REQUIRED } from '@/constants';
import {
  ListingStatus,
  ListingType,
  GlobalUser,
  ProcessedListing,
  RentalStatus,
  SaleStatus,
  FILTER_REASON,
  SORT_ORDER,
} from '@/types';

export async function loadBrandingListings(
  user: GlobalUser,
  listingStatus: ListingStatus
): Promise<ProcessedListing[]> {
  let rentalStatuses: RentalStatus[], saleStatuses: SaleStatus[];

  switch (listingStatus) {
    case ListingStatus.ACTIVE:
      rentalStatuses = [RentalStatus.AVAILABLE];
      saleStatuses = [SaleStatus.AVAILABLE];
      break;
    case ListingStatus.COMING_SOON:
      rentalStatuses = [RentalStatus.COMING_SOON];
      saleStatuses = [SaleStatus.COMING_SOON];
      break;
    case ListingStatus.SOLD:
      rentalStatuses = [RentalStatus.RENTED];
      saleStatuses = [SaleStatus.SOLD];
      break;
    default:
      throw new Error();
  }

  const response = await listSearch({
    /* These are the key parameters: */
    rentalStatuses,
    saleStatuses,
    ownerId: user.personId,
    sortOrder: SORT_ORDER.MOST_RECENT,
    imagesRequired: NUM_IMAGES_REQUIRED,
    /* These parameters are copied over from agent_profile and might need tweaking: */
    listingTypes: [ListingType.LISTING, ListingType.SALES, ListingType.SHORT_TERM_RENTAL],
    includeInContract: true,
    excludePendingSales: false,
    includedFilters: [FILTER_REASON.FURNISHED, FILTER_REASON.KEEP],
    /* These parameters are boring but necessary: */
    rawQuery: '',
    debugLevel: 0,
    start: 0,
    num: 100,
  });

  // Only return listings with media (oddly, `imagesRequired` doesn't do this for us...)
  const listingsWithMedia = response.listings?.filter((listing) => listing.media?.length);
  return listingsWithMedia || [];
}
