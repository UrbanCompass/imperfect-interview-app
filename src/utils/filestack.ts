import * as filestack from 'filestack-js';

import { FILESTACK_API_KEY } from '@/constants';

export const filestackClient = filestack.init(FILESTACK_API_KEY);
