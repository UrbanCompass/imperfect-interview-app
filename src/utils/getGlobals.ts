import { Globals } from '@/types';

import { UC_GLOBALS_ID } from '@/constants';

/* istanbul ignore next */
export function getGlobals(): Globals {
  try {
    const globalsElem = document.getElementById(UC_GLOBALS_ID);
    // eslint-disable-next-line no-eval
    return eval(`(${globalsElem ? globalsElem.textContent : {}})`);
  } catch (e) {
    return {};
  }
}
