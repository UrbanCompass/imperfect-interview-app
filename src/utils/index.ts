export {
  trackAgentBrandingScreenRemoved,
  trackAudioDisabled,
  trackAudioPaused,
  trackAudioPlayed,
  trackAudioSelected,
  trackAutoGenerateClicked,
  trackBackToDigitalAdsClicked,
  trackCreateCustomClicked,
  trackCreateCustomBrandingClicked,
  trackCreateVideoAdClicked,
  trackCompassLogoScreenRemoved,
  trackCustomizationPageViewed,
  trackDownloadClicked,
  trackDownloadPageViewed,
  trackEditVideoClicked,
  trackHomepageViewed,
  trackImageRemoved,
  trackImageReordered,
  trackImageSelected,
  trackImageSelectionPageViewed,
  trackImageUnselected,
  trackImagesUploaded,
  trackImagesSelected,
  trackListingAttributeSelected,
  trackListingSelected,
  trackProgressPageViewed,
  trackStartOverClicked,
  trackVideoCustomized,
} from './analytics';
export { getBaths, getBeds, getListingStatus, getPrice, getSqft } from './listingAttributes';
export { getLogoForAgent } from './logos';
export {
  createMarketingCenterCanvas,
  formatAddressTitle,
  formatAgentDetails,
  formatTextWithLineBreak,
  formatAddressTitleForApiService,
  formatTextForApiService,
  getDigitalAdType,
  getCardType,
  getDefaultAddress,
  getDefaultAgentInfo,
  getInputValue,
  generateMultiLineTextList,
  getTextSource,
  getVideoTitle,
  hasCustomText,
  isAddressScreen,
  isAgentBrandingScreen,
  isCompassLogoScreen,
  isSqftEnabled,
} from './shared';
export { loadBrandingListings } from './branding';
export { filestackClient } from './filestack';
