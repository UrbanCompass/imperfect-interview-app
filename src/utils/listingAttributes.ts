import { formatBathsWithPluralizedLabel, getBathrooms } from '@uc/format-listing-baths';
import { formatBedroomsWithPluralizedLabel } from '@uc/format-listing-bedrooms';
import { formatListingPrice } from '@uc/format-listing-price';

import { ProcessedListing } from '@/types';

export function getBeds(listing: ProcessedListing): string | null {
  return formatBedroomsWithPluralizedLabel(listing, 'Bed' /* suffix */);
}

export function getBaths(listing: ProcessedListing): string | null {
  return formatBathsWithPluralizedLabel(getBathrooms(listing), 'Bath' /* suffix */);
}

export function getSqft(listing: ProcessedListing): string | null {
  const sqft = listing?.size?.squareFeet;
  return sqft ? `${sqft} SQ. FT.` : null;
}

export function getListingStatus(listing: ProcessedListing): string | null {
  return listing?.localizedStatus || null;
}

export function getPrice(listing: ProcessedListing): string | null {
  return formatListingPrice(listing, undefined /*  rentalPeriod */, {
    abbreviateThousands: true,
    abbreviateMillions: true,
  });
}
