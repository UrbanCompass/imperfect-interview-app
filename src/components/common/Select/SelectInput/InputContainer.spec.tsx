import React, { useContext } from 'react';
import { mount, shallow } from 'enzyme';
import { mocked } from 'ts-jest/utils';

import { mockComponent } from '@helpers/mockComponent';
import { InputContainer, Input } from './InputContainer';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn(),
}));

jest.mock('@/components/common/Select', () => ({
  SelectContext: mockComponent('SelectContext'),
}));

test('Input', (): void => {
  expect(mount(<Input />)).toMatchSnapshot();
});

const onValueChange = jest.fn();

describe('InputContainer', (): void => {
  const getInputProps = jest.fn((val) => val);
  const toggleMenu = jest.fn();
  const ctx = {
    placeholder: 'my placeholder',
    getInputProps,
    inputId: 'my-id',
    toggleMenu,
    value: 'foo',
    onValueChange,
    isOpen: true,
  };
  let wrapper;

  describe('default', (): void => {
    beforeAll((): void => {
      mocked(useContext).mockReturnValue(ctx);
      wrapper = shallow(<InputContainer />);
    });

    afterEach(() => {
      toggleMenu.mockClear();
    });

    it('renders default', (): void => {
      expect(wrapper).toMatchSnapshot();
    });

    it('toggles Menu', (): void => {
      const dropdownMenu = wrapper.find({ id: 'my-id' });
      dropdownMenu.simulate('click');
      expect(toggleMenu).toHaveBeenCalledTimes(1);
    });

    it('onValueChange is called if value is set', () => {
      const onKeyDown = wrapper.find('Input').prop('onKeyDown');
      onKeyDown({ key: 'Enter' });
      expect(onValueChange).toHaveBeenCalled();
    });

    it('toggles menu to close on Enter', () => {
      const onKeyDown = wrapper.find('Input').prop('onKeyDown');
      onKeyDown({ key: 'Enter' });
      expect(onValueChange).toHaveBeenCalled();
    });
  });
});
