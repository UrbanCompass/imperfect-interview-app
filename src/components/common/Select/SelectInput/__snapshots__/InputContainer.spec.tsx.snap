// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input 1`] = `
.emotion-0 {
  box-sizing: border-box;
  -webkit-user-select: text;
  -moz-user-select: text;
  -ms-user-select: text;
  user-select: text;
  width: 100%;
}

<Input>
  <input
    className="emotion-0"
  />
</Input>
`;

exports[`InputContainer default renders default 1`] = `
<InputWrapper>
  <Input
    autoCapitalize="off"
    autoComplete="off"
    autoCorrect="off"
    className="cx-selectField-arrow"
    data-tn="tours-my-id-select-input"
    id="my-id"
    onClick={[Function]}
    onKeyDown={[Function]}
    placeholder="my placeholder"
    spellCheck={false}
    type="text"
  />
</InputWrapper>
`;
